### ISSUE TITLE
Clarification on the behaviour of isbounded for "unbounded" HPolygon and HPolytope 

### ISSUE THREAD

**Issue by ueliwechsler** [JuliaReach/LazySets.jl#1914]
I stumbled over this behaviour for an unbounded polyhedral set
```julia
X1 = HPolyhedron([0. 1.], [2.]); isbounded(X1) # false
X2 = HPolygon([0. 1.], [2.]); isbounded(X2) # true
X3 = HPolytope([0. 1.], [2.]); isbounded(X3) # true
```
From the documentation for `HPolygon` we have that:
>  Another type assumption is that the polygon is bounded. The option check_boundedness can  
  be used to assert this. This option is deactivated by default 
... 
>  The function isbounded can be used to manually assert boundedness.

Shouldn't the method `isbounded` return false for the `HPolygon` and `HPolytope`?

**Comment by schillic** [comment]
This is a shortcoming in the documentation. It should say:

> The function `isbounded(P, false)` can be used to manually assert boundedness.

The `false` here ignores the type assumption and actually performs a check.

**Comment by ueliwechsler** [comment]
Thank you! But then, we should also use `isbounded(P, false)` instead of `isbounded(P)` in  #576,  right?

**Comment by schillic** [comment]
No, we do not have to guarantee correct behavior if you use a bounded type to represent an unbounded set. This is a user error.

**Comment by mforets** [comment]
> Shouldn't the method isbounded return false for the HPolygon and HPolytope?

There is always this tension between using only the type information or the value for a `isxyz` function. And I agree that it's unexpected to get `true` in the cases above; after all, someone can argue what is the purpose of `isbounded` if it doesn't check the actual it receives, i.e., its value? There is a kwarg for the "expected" behavior, but still.. A reasonable compromise that I like, and we started to adopt for some newer functions, is to define two functions, `isboundedtype` (uses the value), and `isbounded` (doesn't use the value).

**Comment by ueliwechsler** [comment]
> This is a user error.

I agree, but we could help the user not make this mistake of using a bounded type for an unbounded set. The problem is that the error messages tend to be quite ugly and hard to figure out for these cases since they are not handled as exceptions.

What about warning the user or throwing an error if he tries to create "a bounded unbounded set" such as `HPolytope([0;1.],2.)`? 


> `isbounded` (uses the value), and `isboundedtype` (only uses type information if it is available).

Nice, this would make it more intuitive for me.

**Comment by schillic** [comment]
> What about warning the user or throwing an error if he tries to create "a bounded unbounded set" such as HPolytope([0;1.],2.)?

There is an optional argument `check_boundedness` in the constructor. It is deactivated by default since the check is quite expensive. I am fine with changing activating this check by default. However, we should make sure that we deactivate it explicitly whenever we create an `HPolytope` in library code (because we know that we do not violate this assumption).

**Comment by ueliwechsler** [comment]
> However, we should make sure that we deactivate it explicitly whenever we create an `HPolytope` in library code

I have no idea how much work it would take to do so, but I'd prefer to have `check_boundedness=true` as default for `HPolytope` and `HPolygon` and in general for all bounded types that could be misused to contain an unbounded set (if there are more).

**Comment by mforets** [comment]
> I'd prefer to have check_boundedness=true as default for HPolytope and HPolygon and in general for all bounded types that could be misused

I agree as well to adding more background checks. This helps to reduce the "I stumbled over this behaviour" reaction :) But as @schillic commented, we chose to not activate this check because we haven't worked too much on that aspect to make it relatively efficient: `isbounded(H, false)` if `H::HPolygon` takes ~42us in master. I find that to be too expensive, but if if we can make it below 1us that would be reasonable... I've opened a project, https://github.com/JuliaReach/LazySets.jl/projects/5 and put issues related to boundedness there.
